File: elixir/benchmarking/lib/benchmarking.ex
defmodule Benchmarking do
  @moduledoc """
  Documentation for `Benchmarking`.
  """

  @doc """
  Hello world.

  ## Examples

      iex> Benchmarking.hello()
      :world

  """
  def run_all do
    sqlite()
    IO.puts("")
    postgres()
  end

  def sqlite do
    {:ok, conn} = Exqlite.Sqlite3.open(":memory:")
    :ok = Exqlite.Sqlite3.execute(conn, "create table numbers (name varchar(30), val int)")

    insert_runtime = Benchmarking.measure(fn -> Benchmarking.sqlite_inserts(conn) end)
    IO.puts("sqlite insert runtime (1000 iterations): #{Float.round(insert_runtime * 1000, 3)}ms")

    :ok = Exqlite.Sqlite3.execute(conn, "delete from numbers")

    insert2_runtime = Benchmarking.measure(fn -> Benchmarking.sqlite_inserts2(conn) end)
    IO.puts("sqlite insert2 runtime (1000 iterations): #{Float.round(insert2_runtime * 1000, 3)}ms")

    select_runtime = Benchmarking.measure(fn -> Benchmarking.sqlite_selects(conn) end)
    IO.puts("sqlite select runtime (1000 iterations): #{Float.round(select_runtime * 1000, 3)}ms")

    :ok
  end

  def sqlite_inserts(conn) do
    {:ok, statement} = Exqlite.Sqlite3.prepare(conn, "insert into numbers (name, val) values (?1, ?2)")
    Enum.each((1..1000), fn i ->
      # IO.puts(i)
      :ok = Exqlite.Sqlite3.bind(conn, statement, [Integer.to_string(i), i])
      :done = Exqlite.Sqlite3.step(conn, statement)
    end)
  end

  def sqlite_inserts2(conn) do
    Enum.each((1..1000), fn i ->
      :ok = Exqlite.Sqlite3.execute(conn, "insert into numbers (name, val) values ('#{i}', #{i})")
    end)
  end

  def sqlite_selects(conn) do
    Enum.each((1..1000), fn i ->
      {:ok, statement} = Exqlite.Sqlite3.prepare(conn, "select * from numbers where val = #{i}")
      {:row, _} = Exqlite.Sqlite3.step(conn, statement)
      :done = Exqlite.Sqlite3.step(conn, statement)
    end)
  end

  def postgres do
    {:ok, pid} = Postgrex.start_link(hostname: "db", username: "postgres", password: "password", database: "postgres")
    Postgrex.query!(pid, "drop table if exists numbers", [])
    Postgrex.query!(pid, "create table numbers (name varchar(30), val int)", [])

    insert_runtime = Benchmarking.measure(fn -> Benchmarking.postgres_inserts(pid) end)
    IO.puts("postgres insert runtime (1000 iterations): #{Float.round(insert_runtime * 1000, 3)}ms")

    select_runtime = Benchmarking.measure(fn -> Benchmarking.postgres_selects(pid) end)
    IO.puts("postgres select runtime (1000 iterations): #{Float.round(select_runtime * 1000, 3)}ms")

    IO.puts("end postgres")
  end

  def postgres_inserts(pid) do
    Enum.each((1..1000), fn i ->
      _result = Postgrex.query!(pid, "insert into numbers (name, val) values ($1, $2)", [Integer.to_string(i), i])
    end)
  end

  def postgres_selects(pid) do
    Enum.each((1..1000), fn i ->
      _result = Postgrex.query!(pid, "select * from numbers where val = $1", [i])
    end)
  end

  def measure(function) do
    function
    |> :timer.tc
    |> elem(0)
    |> Kernel./(1_000_000)
  end
end

File: go/benchmarking.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"time"
	_ "github.com/lib/pq"
	_ "github.com/mattn/go-sqlite3"
)

func main() {
	sqliteDb, err := sql.Open("sqlite3", ":memory:")
	if err != nil { log.Fatal(err); return }
	defer sqliteDb.Close()

	pgDb, err := sql.Open("postgres", "host=db user=postgres dbname=postgres password=password sslmode=disable")
	if err != nil { log.Fatal(err); return }
	defer pgDb.Close()

	createTable(sqliteDb)
	createTable(pgDb)

	sqliteInserts(sqliteDb)
	sqliteSelects(sqliteDb)

	fmt.Println("")

	postgresInserts(pgDb)
	postgresSelects(pgDb)
}

func createTable(db *sql.DB) {
	sqlStmt := "create table if not exists numbers (name varchar(30), val int)"
	_, err := db.Exec(sqlStmt)
	if err != nil { log.Printf("%q: %s\n", err, sqlStmt); return }
}

func sqliteInserts(db *sql.DB) {
	defer timeTrack(time.Now(), "sqlite inserts")

	stmt, err := db.Prepare("insert into numbers(name, val) values(?, ?)")
	if err != nil { log.Fatal(err); return }
	defer stmt.Close()

	for i := 0; i < 1000; i++ {
		_, err = stmt.Exec(fmt.Sprintf("%d", i), i)
		if err != nil { log.Fatal(err); return }
	}
}


func postgresInserts(db *sql.DB) {
	defer timeTrack(time.Now(), "postgres inserts")

	stmt, err := db.Prepare("insert into numbers(name, val) values($1, $2)")
	if err != nil { log.Fatal(err); return }
	defer stmt.Close()

	for i := 0; i < 1000; i++ {
		_, err = stmt.Exec(fmt.Sprintf("%d", i), i)
		if err != nil { log.Fatal(err); return }
	}
}

func sqliteSelects(db *sql.DB) {
	defer timeTrack(time.Now(), "sqlite selects")

	stmt, err := db.Prepare("select name from numbers where val = ?")
	if err != nil { log.Fatal(err); return }
	defer stmt.Close()

	var name string
	for i := 0; i < 1000; i++ {
		err = stmt.QueryRow(i).Scan(&name)
		if err != nil { log.Fatal(err); return }
		// fmt.Println(name)
	}
}

func postgresSelects(db *sql.DB) {
	defer timeTrack(time.Now(), "postgres selects")

	stmt, err := db.Prepare("select name from numbers where val = $1")
	if err != nil { log.Fatal(err); return }
	defer stmt.Close()

	var name string
	for i := 0; i < 1000; i++ {
		err = stmt.QueryRow(i).Scan(&name)
		if err != nil { log.Fatal(err); return }
		// fmt.Println(name)
	}
}

func timeTrack(start time.Time, name string) {
	elapsed := time.Since(start)
	log.Printf("%s took %s", name, elapsed)
}
